src/dist/product.rs

```diff
@@ -0,0 +1,401 @@
+use crate::{
+    prelude::DataOrSuffStat,
+    traits::{
+        Cdf, ConjugatePrior, ContinuousDistr, DiscreteDistr, Entropy,
+        HasDensity, HasSuffStat, Mean, Median, Mode, Sampleable, SuffStat,
+        Support,
+    },
+};
+
+macro_rules! tuple_sampleable {
+    ($($len:expr => ($($n:tt $t:ident $x:ident)+))+) => {
+        $(
+            impl<$($t,)+$($x,)+> Sampleable<($($x,)+)> for ($($t,)+)
+            where
+                $($t: Sampleable<$x>,)+
+            {
+                fn draw<R: rand::Rng>(&self, rng: &mut R) -> ($($x,)+) {
+                    (
+                        $(self.$n.draw(rng),)+
+                    )
+                }
+            }
+        )+
+    };
+}
+
+macro_rules! tuple_has_density {
+    ($($len:expr => ($($n:tt $t:ident $x:ident)+))+) => {
+        $(
+            impl<$($t,)+$($x,)+> HasDensity<($($x,)+)> for ($($t,)+)
+            where
+                $($t: HasDensity<$x>,)+
+            {
+                fn ln_f(&self, x: &($($x,)+)) -> f64 {
+                    0.0 $(+ self.$n.ln_f(&x.$n))+
```

Comment:
This can be reworked to omit the zero value by capturing the first element and the rest.